File: solutions/ruby/zebra-puzzle/1/zebra_puzzle.rb
# frozen_string_literal: true

# module ZebraPuzzle
module ZebraPuzzle
  module_function

  HOUSES = (1..5).to_a.freeze
  ORDERS = HOUSES.permutation.to_a.freeze
  FIRST = 1
  MIDDLE = 3
  RESIDENTS = %i[englishman spaniard ukrainian norwegian japanese].freeze

  def water_drinker = solve[:water]
  def zebra_owner = solve[:zebra]
  def adjacent?(a, b) = (a - b).abs == 1

  def solve
    @solve ||= ORDERS.each do |red, green, ivory, yellow, blue|
      next unless green == ivory + 1

      ORDERS.each do |englishman, spaniard, ukrainian, norwegian, japanese|
        next unless englishman == red
        next unless norwegian == FIRST
        next unless adjacent?(norwegian, blue)

        ORDERS.each do |coffee, tea, milk, orange, water|
          next unless coffee == green
          next unless ukrainian == tea
          next unless milk == MIDDLE

          ORDERS.each do |dog, snails, fox, horse, zebra|
            next unless spaniard == dog

            ORDERS.each do |chess, dancing, painter, reading, football|
              next unless japanese == chess
              next unless snails == dancing
              next unless painter == yellow
              next unless football == orange
              next unless adjacent?(reading, fox)
              next unless adjacent?(painter, horse)

              by_house = RESIDENTS.zip([englishman,
                                        spaniard,
                                        ukrainian,
                                        norwegian,
                                        japanese]).to_h.invert

              return {
                water: by_house.fetch(water).to_s.capitalize,
                zebra: by_house.fetch(zebra).to_s.capitalize
              }
            end
          end
        end
      end
    end
  end
end

File: solutions/clojure/zebra-puzzle/1/src/zebra_puzzle.clj
(ns zebra-puzzle
  (:require [clojure.set :as set]))

(defn- permutations [values]
  (if (seq values)
    (for [value values
          remaining (permutations (remove #{value} values))]
      (cons value remaining))
    '(())))

(defn- house-arrangements [attributes] (map #(zipmap attributes %) (permutations (range 5))))
(defn- next-door? [house-a house-b] (= 1 (abs (- house-a house-b))))

(defn- solve []
  (let [resident-arrangements (house-arrangements [:englishman :spaniard :ukrainian :norwegian :japanese])
        color-arrangements (house-arrangements [:blue :red :ivory :green :yellow])
        drink-arrangements (house-arrangements [:tea :coffee :milk :orange :water])
        pet-arrangements (house-arrangements [:dog :snails :horse :fox :zebra])
        hobby-arrangements (house-arrangements [:chess :dancing :painting :reading :football])]
    (first
     (for [resident-houses resident-arrangements
           :when (zero? (:norwegian resident-houses))
           color-houses color-arrangements
           :when (= (inc (:ivory color-houses)) (:green color-houses))
           :when (= (:englishman resident-houses) (:red color-houses))
           :when (next-door? (:norwegian resident-houses) (:blue color-houses))
           drink-houses drink-arrangements
           :when (= 2 (:milk drink-houses))
           :when (= (:coffee drink-houses) (:green color-houses))
           :when (= (:ukrainian resident-houses) (:tea drink-houses))
           pet-houses pet-arrangements
           :when (= (:spaniard resident-houses) (:dog pet-houses))
           hobby-houses hobby-arrangements
           :when (= (:dancing hobby-houses) (:snails pet-houses))
           :when (= (:painting hobby-houses) (:yellow color-houses))
           :when (next-door? (:reading hobby-houses) (:fox pet-houses))
           :when (next-door? (:painting hobby-houses) (:horse pet-houses))
           :when (= (:football hobby-houses) (:orange drink-houses))
           :when (= (:japanese resident-houses) (:chess hobby-houses))
           :let [resident-by-house (set/map-invert resident-houses)]]
       {:water-drinker (resident-by-house (:water drink-houses))
        :zebra-owner (resident-by-house (:zebra pet-houses))}))))

(def ^:private solution (delay (solve)))
(defn drinks-water [] (:water-drinker @solution))
(defn owns-zebra [] (:zebra-owner @solution))
